spring-xd-dirt/src/main/java/org/springframework/xd/dirt/job/dsl/JobSpecification.java

```diff
@@ -929,9 +945,32 @@ private int findNode(String targetJob) {
 					properties.put(arg.getName(), arg.getValue());
 				}
 			}
-			Node node = new Node(Integer.toString(nextId), jr.getName(), null, properties);
-			nodes.add(node);
+			Node node;
+			int nextId;
+
+			// Check if this walk has hit something created earlier in this flow
+			// which may have occurred if a transition referred to it, e.g.
+			// foo | fail=bar || goo || bar
+			// where 'walk'ing foo will have created a node for bar - reuse it.
+
+			Node existingNode = contexts.peek().nodesSharedInFlow.get(jr.getName());
+			if (existingNode != null) {
+				// Found it, reuse it here
+				node = existingNode;
+				nextId = Integer.parseInt(existingNode.id);
+			}
+			else {
+				// Not already in this flow, create a new one
+				nextId = id++;
+				node = new Node(Integer.toString(nextId), jr.getName(), null, properties);
+				nodes.add(node);
+			}
 			createdNodes.put(jr.getName(), node);
+
+			// Should nodes created in the mainline visit be inserted into the shared set?
+			// (In addition to those created for transitions). Do we need it to support
+			// back references (loops?)
```

Comment:
I agree no for right now - left this in more as a reminder to say when someone asks for loops, might need to do something here.